hnn_core/network.py

```diff
@@ -509,8 +554,144 @@ def plot_input(self, ax=None, show=True):
+        Each element of the outer list is a trial.
+        The inner list contains the time stamps of spikes.
+    gids : list (n_trials,) of list (n_spikes,) of float, shape
+        Each element of the outer list is a trial.
+        The inner list contains the cell IDs of neurons that
+        spiked.
+    types : list (n_trials,) of list (n_spikes,) of float, shape
+        Each element of the outer list is a trial.
+        The inner list contains the type of spike (e.g., evprox1
+        or L2_pyramidal) that occured at the corresonding time stamp.
+        Each gid corresponds to a type via Network::gid_dict.
+
+    Methods
+    -------
+    update_types(gid_dict)
+        Update spike types in the current instance of Spikes.
+    plot(ax=None, show=True)
+        Plot and return a matplotlib Figure object showing the
+        aggregate network spiking activity according to cell type.
+    write(fname)
+        Write spiking activity to a collection of spike trial files.
+    '''
+
+    def __init__(self, times=None, gids=None, types=None):
+        if times is None:
+            times = list()
+        if gids is None:
+            gids = list()
+        if types is None:
+            types = list()
+
+        # Validate arguments
+        arg_names = ['times', 'gids', 'types']
+        for arg_idx, arg in enumerate([times, gids, types]):
+            # Validate outer list
+            if not isinstance(arg, list):
+                raise TypeError('%s should be a list of lists'
+                                % (arg_names[arg_idx],))
+            # If arg is not an empty list, validate inner list
+            for trial_list in arg:
+                if not isinstance(trial_list, list):
+                    raise TypeError('%s should be a list of lists'
+                                    % (arg_names[arg_idx],))
+            # Set the length of 'times' as a references and validate
+            # uniform length
+            if arg == times:
+                n_trials = len(times)
+            if len(arg) != n_trials:
+                raise ValueError('times, gids, and types should be lists of '
+                                 'the same length')
+        self._times = times
+        self._gids = gids
+        self._types = types
+
+    def __repr__(self):
+        class_name = self.__class__.__name__
+        n_trials = len(self._times)
+        return '<%s | %d simulation trials>' % (class_name, n_trials)
+
+    def __eq__(self, other):
+        if not isinstance(other, Spikes):
+            return NotImplemented
+        # Round each time element
+        times_self = [[round(time, 3) for time in trial]
+                      for trial in self._times]
+        times_other = [[round(time, 3) for time in trial]
+                       for trial in other._times]
+        return (times_self == times_other and
+                self._gids == other._gids and
+                self._types == other._types)
```

Comment:
I think if you did `np.allclose(times_self, times_other, atol=1e-3) and self._types == other._types` it would do the same thing?

Reply:
I tried it in the beginning but didn't care for the fact that `np.allclose()` uses broadcasting (i.e., it won't control for the case when there has been a formatting discrepancy between `self._times` and `other._times`.

Reply:
humm ... but you won't have any formatting discrepancy now? Because you do input validation.

Reply:
There shouldn't be any formatting discrepancies due to inputs, but if a bug were to emerge in `Network` that results in a formatting discrepancy (e.g., if someone were to mess with some square brackets so that the lists append along the first dimension), it would benefit us to have an `__eq__()` function that is independent. What if we use `np.array_equal()`?

Reply:
ufff right, the Network class. Okay let's leave it like this for now. We can revisit it later.